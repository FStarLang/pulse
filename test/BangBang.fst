module BangBang
open Pulse
#lang-pulse

fn modify_nested_ptr (x: ref (ref Int32.t)) (#y: erased (ref Int32.t)) (#z: erased Int32.t { Int32.v z < Int.max_int Int32.n - 3 })
  preserves x |-> y
  requires reveal y |-> z
  returns z': Int32.t
  ensures reveal y |-> Int32.add z 2l
  ensures rewrites_to z' (Int32.add z 3l)
{
  (!x) := (!(!x)) `Int32.add` 2l;
  ((!(!x)) `Int32.add` 1l)
}

fn decr_uint () {
  let mut x: UInt32.t = 1ul;
  x := !x `UInt32.sub` 1ul;
}
