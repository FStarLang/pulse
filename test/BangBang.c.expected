/* krml header omitted for test repeatability */


#include "BangBang.h"

int32_t BangBang_modify_nested_ptr(int32_t **x)
{
  int32_t *__anf2737 = *x;
  int32_t *__anf2733 = *x;
  int32_t __anf2735 = *__anf2733;
  *__anf2737 = __anf2735 + (int32_t)2;
  int32_t *__anf3458 = *x;
  int32_t __anf3460 = *__anf3458;
  return __anf3460 + (int32_t)1;
}

void BangBang_decr_uint(void)
{
  uint32_t x = 1U;
  uint32_t __anf136 = x;
  x = __anf136 - 1U;
}

